### PR TITLE
feat: allow caching and resuming picker

### LINKED ISSUE #751: resume previous list
**Is your feature request related to a problem? Please describe.**
I'm switching from [denite](https://github.com/Shougo/denite.nvim), and the one thing I'm missing is its `-resume` flag, which simply restores the previous list, in the same filtered state/order. This allows me to, for example, search my project for a token, then go through the list and make modifications. The quickfix workflow from https://github.com/nvim-telescope/telescope.nvim/issues/52 is close, but it doesn't give me the ability to keep track of where in the list my current progress is.

**Describe the solution you'd like**
`:Telescope resume` to simply re-open my last Telescope search, in the same state.
`:Telescope resume cursor_pos=+1` to re-open my last Telescope search with the cursor one line above the previous state.
`:Telescope resume cursor_pos=-1` to re-open my last Telescope search with the cursor one line below the previous state.

**Describe alternatives you've considered**
`:Telescope next/previous` to take the default action on the next/previous item in the previous search, e.g. open the next buffer after `:Telescope buffers`.

**Additional context**
https://github.com/Shougo/denite.nvim/blob/master/doc/denite.txt#L905
https://github.com/Shougo/denite.nvim/blob/master/doc/denite.txt#L684

### PR DESCRIPTION
**Pull request by fdschmidt93**
Will close #751

What?
- Introduce `resume` (resume last picker) and `pickers` (picker over cached pickers) `builtin` pickers

How?
- Modify `pickers.lua`  so picker attributes are not `nil`-ed on detach but stored in global `cached_pickers` table-stack instead
- `cache_limit` > 0 reduces # of cached entries via `LinkedList:truncate` (primarily to reduce size)
- New buffer previewer that writes original results in preview

Showcase

https://user-images.githubusercontent.com/39233597/127784598-e3d32413-2957-4fc5-8a1e-8ea6598b42b2.mp4

### CODE REVIEW

File: lua/telescope/pickers.lua
Comment (fdschmidt93) [inline lua/telescope/pickers.lua:432]: I've been dreading already too long over this.

The straightforward approach of 

```lua
local index = 1
for entry in self.manager:iter() do
  self:entry_adder(index, entry, _, true)
  index = index + 1
  if self:can_select_row(row) and self._multi:is_selected(entry) then
    self.highlighter:hi_multiselect(row, true)
  end
end
```

doesn't work. I've tried a lot of other hacks I could think of but to no avail. In any case, selected entries in any case would be highlighted once you scroll over them.

File: lua/telescope/pickers.lua
Comment (fdschmidt93) [inline lua/telescope/pickers.lua:432]: I've been dreading already too long over this.

The straightforward approach of 

```lua
local index = 1
for entry in self.manager:iter() do
  self:entry_adder(index, entry, _, true)
  index = index + 1
  if self:can_select_row(row) and self._multi:is_selected(entry) then
    self.highlighter:hi_multiselect(row, true)
  end
end
```

doesn't work. I've tried a lot of other hacks I could think of but to no avail. In any case, selected entries in any case would be highlighted once you scroll over them.

File: lua/telescope/algos/linked_list.lua
Comment (tjdevries) [inline lua/telescope/algos/linked_list.lua:218]: Hmm, we could decide whether to go from the front or the end based on the length and the max results.

For example, if we have 500,000 results and we limit 700,000 then there's actually no reason to go at all.

If we have 1,000,001 results and limit is 1,000,000 better to just truncate the last entry.

File: doc/telescope.txt
Comment (tjdevries) [inline doc/telescope.txt:35]: I think our default could be even smaller if it affects performance at all.

**Comment by fdschmidt93** [comment]
I'd be to get advice on whether the buffer previewer can be done more cleverly to universally account for (filtered) ascending or descending results. Lines apparently cannot be placed at an arbitrary row for empty buffers.

File: doc/telescope.txt
Comment (fdschmidt93) [reply doc/telescope.txt]: Agreed

**Comment by tami5** [comment]
> It would be great to persist pickers across sessions by writing them to disk as we then can preserve multi selections for cool stuff.
> 
> 
> 
> However, it quite challenging:
> 
> - Pickers across projects are mostly irrelevant to one another -- requires likely something like git root resolved workspace storage
> 
> - Serializing lua tables appears tedious for that use case ... really happy to hear ideas on this as I think it's super cool :)

I'd say use sql.nvim, it is fully equipped with this kind of stuff 🤣. But @tjdevries hate being tide to yet another dependency. Alternatively, look at how I serialized and deserialized lua tables, it's simply. then maybe save and load then from a file.

btw great job @fdschmidt93

**Comment by tjdevries** [comment]
@fdschmidt93 I think saving to disk would make a good extension :) (as @tami5 would predict that I would say). If we end up finding simple, OS agnostic, performant solutions over time there, then we can talk about moving it back into core, but I do not want to hold up a PR and/or create undue complexity for that feature at this time.

Additionally, if you make it an extension, you can require sql.nvim, or other luarocks packages if desired.

**Comment by tjdevries** [comment]
I do not like that the pickers are cached forever. We should make a maximum size so that we release the pickers at some point in the future (for example, another config option is OK to do perhaps). We should set this to something small by default, perhaps only 1.

This will prevent keeping around 100,000 entries for the rest of your nvim editing session, which would be very confusing for people who just boot up telescope for the first time and live grep somewhere a few times.

**Comment by tjdevries** [comment]
Also, if we add more entries, then I would move us towards doing something like having a `resume` option that takes several configuration options for pickers. Idk, I'd have to think a bit more about that generally then.

**Comment by fdschmidt93** [comment]
Thanks @tami5 :)

I've briefly looked at `sql.nvim/json.lua` and if I understand all correctly it's not look I cannot easily dump a complex table like a picker and its finders, entries, and selections in there without significant wrappers or telescope.nvim changes to make my life easier. As in, the way to probably best use `sql.nvim` would be to parse entries, selections, and whatever required to reinstantiate the picker and then run `find`. Coming from Python, I was looking for something like `pickle` :sweat_smile: 

@tjdevries 

Agreed on the extension :)

Makes sense, I'll add a limit of # cached pickers and expose it as a config. For context, as is, it's not really an issue since the entries are truncated to the first 1K by default.

**Comment by tjdevries** [comment]
I can't really imagine a usecase for looking at the 327th picker ago :laughing:

**Comment by fdschmidt93** [comment]
We now
- Cache only `cache_n_pickers` (defaullt: 10, @tjdevries if you feel strongly about only the last, happy to change though from all my tests it should be fine)
- Can override `opts` of `builtin.resume` and `builtin.pickers` (latter passed onto `resume` ofc) consistent with other pickers

:)

File: lua/telescope/builtin/internal.lua
Comment (Conni2461) [inline lua/telescope/builtin/internal.lua:4]: Not needed we already have `conf` imported (unless you need something else from config, doesn't look like it rn)

File: lua/telescope/config.lua
Comment (Conni2461) [inline lua/telescope/config.lua:282]: Rather than having 3 (or more) values could we do something similar to `layout_config` and `history` where we have one table setting with multiple keys?

File: lua/telescope/builtin/internal.lua
Comment (fdschmidt93) [reply lua/telescope/builtin/internal.lua:4]: Yes, slight oversight sorry :) fixed locally and to be updated once more comes in

File: lua/telescope/pickers.lua
Comment (Conni2461) [inline lua/telescope/pickers.lua:533]: Is this a workaround for opening a picker in an action if telescope wasn't closed with `actions._close(prompt_bufnr, true)`

File: lua/telescope/builtin/internal.lua
Comment (Conni2461) [inline lua/telescope/builtin/internal.lua:118]: You should close with `actions._close(prompt_bufnr, true)`

File: lua/telescope/pickers.lua
Comment (fdschmidt93) [reply lua/telescope/pickers.lua:533]: Sort of, yes 😅 as I didn't recall `_close` vs `close` in `builtin.internal.pickers`.

Not sure whether to keep this (since it shouldn't hurt?) or only change to `_close` in pickers action.

File: lua/telescope/previewers/buffer_previewer.lua
Comment (Conni2461) [inline lua/telescope/previewers/buffer_previewer.lua:892]: Can we flip this to make it a little bit more readable. `if self.state.bufname then return end`. I currently having a hard time with the indents and all the ends :laughing:

File: lua/telescope/config.lua
Comment (fdschmidt93) [reply lua/telescope/config.lua:293]: I hope this is fine now; I'm afraid this increased control flow complexity quite a bit as I'm writing to the `pickers.cache_picker` during caching and `opts.cache_picker = false` can be validly passed to `builtin.{resume, pickers}`.

**Comment by sodiumjoe** [comment]
Trying out this branch, a couple of notes few notes:
* it would be great if it would also resume the selection state
* it would be great if it would also resume the input buffer mode

The workflow i'm thinking of is quickfix-ish, i.e. I grep for something, then go through the list of results making edits. So when I resume, I want to be in normal mode, with the same selection highlighted.

**Comment by fdschmidt93** [comment]
Many thanks for trying out the branch and the valuable feedback :) I've incorporated your suggestions as far as possible right now.

There still are a couple of issues around initial normal mode, most prominently wrt caching `default_text` does not play nicely and results are only shown once insert mode was entered. I guess most, if not all, issues around this will be fixed in #987 

I'll most likely have to rebase this branch then to #987 anyways and will validate then that these cases around caching work as expected.

Thanks again!

**Comment by sodiumjoe** [comment]
Thanks, this is great! Now that I think about it, I reported https://github.com/nvim-telescope/telescope.nvim/issues/750 which might have something to do with the issues starting in normal mode. I'll continue testing this branch, but so far so good!

File: lua/telescope/config.lua
Comment (tjdevries) [inline lua/telescope/config.lua:295]: I really think this should be 1 by default. If users are interested in storing lots of em, they can do that.

Otherwise, this lets them do the most common use case, which is resuming the last thing they did. I don't think this will happen often to store a bunch of them.

File: lua/telescope/pickers.lua
Comment (tjdevries) [inline lua/telescope/pickers.lua:425]: Hmmm, this section makes me a bit nervous.

We have a few vim.schedules, which I'm hoping we should not have to do necessarily.

Also, with newest update (you'll have to pull from master again) you should be able to do something like await_scheduler() instead, if you really need to do scheduling to make this work out OK.

Also, I'd see if we can re-use more code here or abstract it into some helper function within picker so that we don't share too much state here accidentally.

Does that make sense?

File: lua/telescope/previewers/buffer_previewer.lua
Comment (tjdevries) [inline lua/telescope/previewers/buffer_previewer.lua:238]: I can't find anywhere that this is being used (on first search, just in github browser)

**Comment by tjdevries** [comment]
Other than a few nitpicks, I just want to make sure that this still works with the newest async updates and maybe it will allow writing some of the code a bit easier and more straightforward in the pickers.lua section.

Once we do that, I think we can probably merge!

File: lua/telescope/previewers/buffer_previewer.lua
Comment (Conni2461) [reply lua/telescope/previewers/buffer_previewer.lua:238]: getting a new table ref here should be good. I think we can ignore that comment :laughing:

File: lua/telescope/pickers.lua
Comment (fdschmidt93) [reply lua/telescope/pickers.lua:440]: > re-use more code here

The function for highlighting was already there :laughing: I just hadn't seen it.

Yes, `await_scheduler` is required to make this work.

**Comment by fdschmidt93** [comment]
I've rebased to latest master and addressed the comments. One thing that doesn't work (should it? also seems to affect master), is setting `default_text` for a picker if `initial_mode="normal"`, such that the `default_text` is shown in the prompt.

@sodiumjoe given your experience with this branch, could you also please retest it and report whether everything fine is on your end? If you are using the `pickers` picker then please set caching in your set up adequately, like so:

```lua
telescope.setup {
  defaults = {
    cache_picker = {
      num_pickers = 20 -- your preferred number here, values up to 100 should be perfectly fine; likely even much higher
    }
  }
}
```

File: doc/telescope.txt
Comment (Conni2461) [inline doc/telescope.txt:45]: docs no longer align.

Comment (fdschmidt93) [reply doc/telescope.txt]: Sorry & thanks, fixed.

File: lua/telescope/config.lua
Comment (Conni2461) [reply lua/telescope/config.lua:300]: I like the config now. I am more concerned about a better user interface than a couple more ifs in code :laughing: 
Also i love this table config to group stuff since clason proposed it for history. way better than having multiple string values that are prefixed with something

**Review by Conni2461 — APPROVED** [review]
LGTM

Thanks :) and good job

**Comment by fdschmidt93** [comment]
Thanks :)

**Comment by max397574** [comment]
afaict it isn't possible but ig you could opena feature request issue for this
it sounds like something really useful @yyy33

**Comment by yyy33** [comment]
> afaict it isn't possible but ig you could opena feature request issue for this it sounds like something really useful @yyy33

thank's , i opened it

**Pull request merged**
